python/deltalake/table.py

```diff
@@ -617,25 +624,38 @@ def merge(
         invariants = self.schema().invariants
         checker = _DeltaDataChecker(invariants)
 
+        from .schema import (
+            convert_pyarrow_dataset,
+            convert_pyarrow_recordbatch,
+            convert_pyarrow_recordbatchreader,
+            convert_pyarrow_table,
+        )
+
         if isinstance(source, pyarrow.RecordBatchReader):
-            schema = source.schema
+            source = convert_pyarrow_recordbatchreader(source, large_dtypes=False)
         elif isinstance(source, pyarrow.RecordBatch):
-            schema = source.schema
-            source = [source]
+            source = convert_pyarrow_recordbatch(
+                source, large_dtypes=False
+            )  # TODO(ion): set to True once MERGE uses logical plan
```

Comment:
logical plans for MERGE are merged, do you we want to update this?

Reply:
Yes forgot about this one, will set it to large.

Just a quick note, I noticed the rust engine writer is more flexible in large and normal types, so my thought is to deprecate the part of the schema conversion from large to normal and vice versa, once the python writer uses the rust engine exclusively.